File: api/netdata.py
# Python
import json

# GroundSeg modules
from log import Log
from netdata_docker import NetdataDocker

class Netdata:
    data = {}
    updater_info = {}
    default_config = {
            "netdata_name": "netdata",
            "repo": "registry.hub.docker.com/netdata/netdata",
            "netdata_version": "latest",
            "amd64_sha256": "95e74c36f15091bcd7983ee162248f1f91c21207c235fce6b0d6f8ed9a11732a",
            "arm64_sha256": "cd3dc9d182a4561b162f03c6986f4647bbb704f8e7e4872ee0611b1b9e86e1b0",
            "volume_dir": "/var/lib/docker/volumes",
            "cap_add": ["SYS_PTRACE"],
            "port": 19999,
            "restart": "unless-stopped",
            "security_opt": "apparmor=unconfined",
            "volumes": [
                "netdataconfig:/etc/netdata",
                "netdatalib:/var/lib/netdata",
                "netdatacache:/var/cache/netdata",
                "/etc/passwd:/host/etc/passwd:ro",
                "/etc/group:/host/etc/group:ro",
                "/proc:/host/proc:ro",
                "/sys:/host/sys:ro",
                "/etc/os-release:/host/etc/os-release:ro"
                ]
            }

    def __init__(self, config):
        self.config_object = config
        self.config = config.config
        self.nd_docker = NetdataDocker()
        self.filename = f"{self.config_object.base_path}/settings/netdata.json"

        # Set Netdata Config
        self.load_config()
        branch = self.config['updateBranch']
        self.data = {**self.default_config, **self.data}

        # Updater Netdata information
        if (self.config_object.update_avail) and (self.config['updateMode'] == 'auto'):
            Log.log("Netdata: Replacing local data with version server data")
            self.updater_info = self.config_object.update_payload['groundseg'][branch]['netdata']
            self.data['repo'] = self.updater_info['repo']
            self.data['netdata_version'] = self.updater_info['tag']
            self.data['amd64_sha256'] = self.updater_info['amd64_sha256']
            self.data['arm64_sha256'] = self.updater_info['arm64_sha256']

        # image replaced by repo
        if 'image' in self.data:
            self.data.pop('image')

        self.save_config()

        if self.start():
            Log.log("Netdata: Initialization Completed")

    def start(self):
        return self.nd_docker.start(self.data, self.config_object._arch)

    # Container logs
    def logs(self):
        return self.nd_docker.full_logs(self.data['netdata_name'])

    #def stop(self):
    #    return self.nd_docker.stop()

    # Load netdata.json
    def load_config(self):
        try:
            with open(self.filename) as f:
                self.data = json.load(f)
                Log.log("Netdata: Successfully loaded netdata.json")

        except Exception as e:
            Log.log(f"Netdata: Failed to open netdata.json: {e}")
            Log.log("Netdata: New netdata.json will be created")

    # Save netdata.json
    def save_config(self):
        with open(self.filename,'w') as f:
            json.dump(self.data, f, indent=4)
            f.close()

File: goseg/docker/netdata.go
package docker

import (
	"context"
	"fmt"
	"groundseg/config"
	"groundseg/dockerclient"
	"io/ioutil"
	"os"
	"path/filepath"

	"github.com/docker/docker/api/types"
	"github.com/docker/docker/api/types/container"
	"github.com/docker/go-connections/nat"
	"go.uber.org/zap"
)

func LoadNetdata() error {
	zap.L().Info("Loading NetData container")
	confPath := filepath.Join(config.BasePath, "settings", "netdata.json")
	_, err := os.Open(confPath)
	if err != nil {
		// create a default if it doesn't exist
		err = config.CreateDefaultNetdataConf()
		if err != nil {
			// panic if we can't create it
			errmsg := fmt.Sprintf("Unable to create NetData config! %v", err)
			zap.L().Error(errmsg)
			panic(errmsg)
		}
	}
	err = WriteNDConf()
	if err != nil {
		return err
	}
	zap.L().Info("Running NetData")
	info, err := StartContainer("netdata", "netdata")
	if err != nil {
		zap.L().Error(fmt.Sprintf("Error starting NetData: %v", err))
		return err
	}
	config.UpdateContainerState("netdata", info)
	return nil
}

// netdata container config builder
func netdataContainerConf() (container.Config, container.HostConfig, error) {
	var containerConfig container.Config
	var hostConfig container.HostConfig
	// construct the container metadata from version server info
	containerInfo, err := GetLatestContainerInfo("netdata")
	if err != nil {
		return containerConfig, hostConfig, err
	}
	desiredImage := fmt.Sprintf("%s:%s@sha256:%s", containerInfo["repo"], containerInfo["tag"], containerInfo["hash"])
	// construct the container config struct
	containerConfig = container.Config{
		Image:        desiredImage,
		ExposedPorts: nat.PortSet{"19999/tcp": struct{}{}},
		Volumes: map[string]struct{}{
			"/etc/netdata":         {},
			"/var/lib/netdata":     {},
			"/var/cache/netdata":   {},
			"/host/etc/passwd":     {},
			"/host/etc/group":      {},
			"/host/proc":           {},
			"/host/sys":            {},
			"/host/etc/os-release": {},
		},
	}
	hostConfig = container.HostConfig{
		CapAdd: []string{"SYS_PTRACE"},
		RestartPolicy: container.RestartPolicy{
			Name: "unless-stopped",
		},
		Resources: container.Resources{
			NanoCPUs: 1e8,               // 1cpu
			Memory:   200 * 1024 * 1024, // 200mb
		},
		SecurityOpt: []string{"apparmor=unconfined"},
		PortBindings: nat.PortMap{
			"19999/tcp": []nat.PortBinding{
				{HostIP: "0.0.0.0", HostPort: "19999"},
			},
		},
		Binds: []string{
			"netdataconfig:/etc/netdata",
			"netdatalib:/var/lib/netdata",
			"netdatacache:/var/cache/netdata",
			"/etc/passwd:/host/etc/passwd:ro",
			"/etc/group:/host/etc/group:ro",
			"/proc:/host/proc:ro",
			"/sys:/host/sys:ro",
			"/etc/os-release:/host/etc/os-release:ro",
		},
	}
	return containerConfig, hostConfig, nil
}

// write edited conf
func WriteNDConf() error {
	newConf := "[plugins]\n     apps = no\n"
	filePath := filepath.Join(config.DockerDir, "netdataconfig", "_data", "netdata.conf")
	existingConf, err := ioutil.ReadFile(filePath)
	if err != nil {
		// assume it doesn't exist, so write the current config
		zap.L().Info("Creating ND config")
		return writeNDConfToFile(filePath, newConf)
	}
	if string(existingConf) != newConf {
		// If they differ, overwrite
		zap.L().Info("Writing ND config")
		return writeNDConfToFile(filePath, newConf)
	}
	return nil
}

// either write directly or create volumes
func writeNDConfToFile(filePath string, content string) error {
	// try writing
	err := ioutil.WriteFile(filePath, []byte(content), 0644)
	if err == nil {
		return nil
	}
	// ensure the directory structure exists
	dir := filepath.Dir(filePath)
	if err = os.MkdirAll(dir, 0755); err != nil {
		return err
	}
	// try writing again
	err = ioutil.WriteFile(filePath, []byte(content), 0644)
	if err != nil {
		err = copyNDFileToVolume(filePath, "/etc/netdata/", "netdata")
		// otherwise create the volume
		if err != nil {
			return fmt.Errorf("Failed to copy ND config file to volume: %v", err)
		}
	}
	return nil
}

// write ND conf to volume
func copyNDFileToVolume(filePath string, targetPath string, volumeName string) error {
	ctx := context.Background()
	cli, err := dockerclient.New()
	if err != nil {
		return err
	}
	defer cli.Close()
	containerInfo, err := GetLatestContainerInfo("netdata")
	if err != nil {
		return err
	}
	desiredImage := fmt.Sprintf("%s:%s@sha256:%s", containerInfo["repo"], containerInfo["tag"], containerInfo["hash"])
	// temp container to mount
	resp, err := cli.ContainerCreate(ctx, &container.Config{
		Image: desiredImage,
	}, &container.HostConfig{
		Binds: []string{volumeName + ":" + targetPath},
	}, nil, nil, "nd_writer")
	if err != nil {
		return err
	}
	if err := cli.ContainerStart(ctx, resp.ID, types.ContainerStartOptions{}); err != nil {
		return err
	}
	file, err := os.Open(filepath.Join(filePath))
	if err != nil {
		return fmt.Errorf("failed to open netdata.config file: %v", err)
	}
	defer file.Close()
	// Copy the file to the volume via the temporary container
	err = cli.CopyToContainer(ctx, resp.ID, targetPath, file, types.CopyToContainerOptions{})
	if err != nil {
		return err
	}
	// remove temporary container
	if err := StopContainerByName("nd_writer"); err != nil {
		return err
	}
	if err := cli.ContainerRemove(ctx, resp.ID, types.ContainerRemoveOptions{Force: true}); err != nil {
		return err
	}
	defer func() {
		if removeErr := cli.ContainerRemove(ctx, resp.ID, types.ContainerRemoveOptions{Force: true}); removeErr != nil {
			zap.L().Error(fmt.Sprintf("Failed to remove temporary container: ", removeErr))
		}
	}()
	return nil
}
